File: src/test/java/com/raychenon/leetcode/tree/BinaryTreeLevelOrderTraversalTest.java
package com.raychenon.leetcode.tree;

import org.junit.BeforeClass;
import org.junit.Test;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import static junit.framework.TestCase.assertEquals;

/**
 * User: raychenon
 * Date: 11/9/2020
 * https://leetcode.com/problems/binary-tree-level-order-traversal/
 */
public class BinaryTreeLevelOrderTraversalTest {

    private static BinaryTreeLevelOrderTraversal binaryTreeLevelOrderTraversal;

    @BeforeClass
    public static void setUp() {
        binaryTreeLevelOrderTraversal = new BinaryTreeLevelOrderTraversal();
    }

    /**
     * --- 3
     * -- / \
     * - 9   20
     * ---\  / \
     * 15   7
     */
    @Test
    public void example1Test() {
        TreeNode root = TreeUtil.createTree(Arrays.asList(3, 9, 20, null, null, 15, 7));
        List<List<Integer>> expected = List.of(List.of(3), List.of(9, 20), List.of(15, 7));
        assertMethods(expected, root);
    }

    /**
     * --- 4
     * -- / \
     * - 9   0
     * -/ \
     * 5  1
     */
    @Test
    public void example2Test() {
        TreeNode root = TreeUtil.createTree(Arrays.asList(4, 9, 0, 5, 1));
        List<List<Integer>> expected = List.of(List.of(4), List.of(9, 0), List.of(5, 1));
        assertMethods(expected, root);
    }

    @Test
    public void nullRootTest() {
        TreeNode root = null;
        List<List<Integer>> expected = new LinkedList<>();
        assertMethods(expected, root);
    }


    private void assertMethods(List<List<Integer>> expected, TreeNode root) {
        List<List<Integer>> answer = binaryTreeLevelOrderTraversal.levelOrder(root);
        List<List<Integer>> answerRec = binaryTreeLevelOrderTraversal.levelOrderRec(root);
        for (int i = 0; i < expected.size(); i++) {
            for (int j = 0; j < expected.get(i).size(); j++) {
                assertEquals(expected.get(i).get(j), answer.get(i).get(j));
                assertEquals(expected.get(i).get(j), answerRec.get(i).get(j));
            }
        }
    }
}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/tree/BinaryTreeLevelOrderTraversalTest.kt
package com.raychenon.kotlin.leetcode.tree

import com.raychenon.leetcode.tree.TreeNode
import com.raychenon.leetcode.tree.TreeUtil
import junit.framework.TestCase
import org.junit.Test

/**
 * User: raychenon
 * Date: 9/9/2020
 * https://leetcode.com/problems/binary-tree-level-order-traversal/
 */
class BinaryTreeLevelOrderTraversalTest {

    /**
     * --- 3
     * -- / \
     * - 9   20
     * ---\  / \
     *     15   7
     */
    @Test
    fun example1Test() {
        val root = TreeUtil.createTree(listOf(3, 9, 20, null, null, 15, 7))
        val expected = listOf(listOf(3), listOf(9, 20), listOf(15, 7))
        assertMethods(expected, root)
    }

    /**
     * --- 4
     * -- / \
     * - 9   0
     * -/ \
     *  5  1
     */
    @Test
    fun example2Test() {
        val root = TreeUtil.createTree(listOf(4, 9, 0, 5, 1))
        val expected = listOf(listOf(4), listOf(9, 0), listOf(5, 1))
        assertMethods(expected, root)
    }

    @Test
    fun nullNodeTest() {
        val root = null
        val expected = emptyList<List<Int>>()
        assertMethods(expected, root)
    }

    private fun assertMethods(expected: List<List<Int>>, root: TreeNode?): Unit {
        val answer = BinaryTreeLevelOrderTraversal.levelOrder(root)
        val answerRec = BinaryTreeLevelOrderTraversal.levelOrderRec(root)
        for (i in 0 until expected.size) {
            for (j in 0 until expected[i].size) {
                TestCase.assertEquals(expected[i][j], answer[i][j])
                TestCase.assertEquals(expected[i][j], answerRec[i][j])
            }
        }
    }

}